### PR TITLE
Do not break pipeline when window queries involve GROUPS

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#5586](https://github.com/apache/arrow-datafusion/issues/5586).

# Rationale for this change

Currently, when window queries involve `GROUPS` window frame type. We cannot run it using `BoundedWindowAggExec`. Instead we use `WindowAggExec`. This breaks pipeline, even if there is no theoretical reason to do so. With this change we solve this problem. See the [#5586](https://github.com/apache/arrow-datafusion/issues/5586) for better explanation.

# What changes are included in this PR?

With this PR physical plan of the query
```sql
SELECT
    SUM(c6) OVER(ORDER BY c2 GROUPS BETWEEN 1 PRECEDING AND 1 FOLLOWING) as sum1
    FROM aggregate_test_100
```
would turn from
```sql
"ProjectionExec: expr=[SUM(aggregate_test_100.c6) ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] GROUPS BETWEEN 1 PRECEDING AND 1 FOLLOWING@13 as sum1]",
"  WindowAggExec: wdw=[SUM(aggregate_test_100.c6): Ok(Field { name: \"SUM(aggregate_test_100.c6)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Groups, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
"    SortExec: expr=[c2@1 ASC NULLS LAST]",
"      CsvExec: files={1 group: [[file_path]]}, has_header=true, limit=None, projection=[c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13]",
```
to
```sql
"ProjectionExec: expr=[SUM(aggregate_test_100.c6) ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] GROUPS BETWEEN 1 PRECEDING AND 1 FOLLOWING@13 as sum1]",
"  BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c6): Ok(Field { name: \"SUM(aggregate_test_100.c6)\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Groups, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }]",
"    SortExec: expr=[c2@1 ASC NULLS LAST]",
"      CsvExec: files={1 group: [[file_path]]}, has_header=true, limit=None, projection=[c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13]",
```
 This enables us to not break the pipeline when window queries involve `GROUPS` clause (given that there is no other pipeline-breaking executor in the plan). 

# Are these changes tested?
Yes. Fuzzy tests are extended to cover this use case. Also end to end tests are added.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/window/window_expr.rs [hunk 1]
```diff
@@ -353,6 +371,59 @@ pub struct WindowAggState {
...
+            Some(WindowFrameContext::Groups { state, .. }) => {
+                let mut n_group_to_del = 0;
+                state.group_end_indices.retain_mut(|(_, start_idx)| {
+                    if n_prune >= *start_idx {
+                        n_group_to_del += 1;
+                        false
+                    } else {
+                        true
+                    }
+                });
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_expr.rs:401]: `group_end_indices` seems to have increasing end indices, isn't it? In that case, you can break out of the this loop early if you don't use `retain_mut`.

File: datafusion/physical-expr/src/window/window_expr.rs [hunk 2]
```diff
@@ -340,6 +357,7 @@ pub enum BuiltinWindowState {
 pub struct WindowAggState {
     /// The range that we calculate the window function
     pub window_frame_range: Range<usize>,
+    pub window_frame_ctx: Option<WindowFrameContext>,
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_expr.rs:360]: Is there any information overlap between the attributes `window_frame_range` and what is in `window_frame_ctx`? If so, maybe we can think of a simplifying refactor. If not, please explain to me in our next meeting.

**Review by ozankabak — COMMENTED** [review]
I went through this code carefully and simplified it as best as I can. You were cloning the window state context every time, which was entirely avoidable. I Also switched to a few idiomatic APIs like `get_or_insert_with` to simplify code flow. One change whose correctness I'm not sure of is that I made `prune_state` a no-op if no window state context exists yet. Please check the correctness of this change.

If you have any improvements/refactors in mind, please go ahead work on them and we can discuss everything again and do a second review if necessary.

**Review by ozankabak — APPROVED** [review]
There are only two items remaining here:
1. Writing an explanatory PR body.
2. Investigating the possible removal of duplicate `last_range` from range state.

After these are completed, this will be ready to send upstream.

**Comment by mustafasrepo** [comment]
Merged to upstream

**Pull request closed without merging**